include/swift/AST/DiagnosticsSema.def

```diff
@@ -2972,18 +2972,18 @@ NOTE(protocol_witness_missing_differentiable_attr,none,
 // @derivative
 ERROR(derivative_attr_expected_result_tuple,none,
       "'@derivative(of:)' attribute requires function to return a two-element "
-      "tuple of type '(value: T..., pullback: (U.TangentVector) -> T.TangentVector...)' "
-      "or '(value: T..., differential: (T.TangentVector...) -> U.TangentVector)'", ())
+      "tuple; first element must have label 'value:' and second element must "
+      "have label 'pullback:' or 'differential:'", ())
 ERROR(derivative_attr_invalid_result_tuple_value_label,none,
       "'@derivative(of:)' attribute requires function to return a two-element "
-      "tuple (first element must have label 'value:')", ())
+      "tuple; first element must have label 'value:'", ())
 ERROR(derivative_attr_invalid_result_tuple_func_label,none,
       "'@derivative(of:)' attribute requires function to return a two-element "
-      "tuple (second element must have label 'pullback:' or 'differential:')",
+      "tuple; second element must have label 'pullback:' or 'differential:'",
       ())
 ERROR(derivative_attr_result_value_not_differentiable,none,
       "'@derivative(of:)' attribute requires function to return a two-element "
-      "tuple (first element type %0 must conform to 'Differentiable')", (Type))
+      "tuple; first element type %0 must conform to 'Differentiable'", (Type))
```

Comment:
Why's "conform to `Differentiable`" special here in a derivative function? Would it be better to just emit a diagnostic that suggests the expected derivative type?

Reply:
> Would it be better to just emit a diagnostic that suggests the expected derivative type?

That would be ideal. However, for `@derivative` attribute type-checking, the "expected derivative type" is not known - we start with the type of the `@derivative` declaration and try to compute the appropriate original function type.

Reply:
Okay, but I feel the "must conform to `Differentiable`" error message is making it worse because it seems to suggest that derivative function's `value:` result is somehow special with regards to `Differentiable` conformance, while it's in fact simply the same as the original function's result. Just my two cents.  I think a better diagnostic could be:
> first element type %0 must be the same as the result of the original function that's being differentiated

Reply:
I understand your point. I pushed some further diagnostics improvements along this direction.

Now, original function lookup occurs before the "does first element type conform to `Differentiable`?" check. In effect, this delays "find `Differentiable` conformance for `value:` result" as late as possible, so original function lookup diagnostics trigger first.

---

Example:

```swift
import _Differentiation
@derivative(of: nonexistentFunction)
func derivative(_ x: Float) -> (value: Int, pullback: (Float) -> Float) {
  fatalError()
}
```

Before: unideal diagnostic about `value:` result.
```
derivative.swift:2:2: error: '@derivative(of:)' attribute requires function to return a two-element tuple (first element type 'Int' must conform to 'Differentiable')
@derivative(of: nonexistentFunction)
 ^
```

After: ideal diagnostic about original function not found.
```
derivative.swift:2:17: error: use of unresolved identifier 'nonexistentFunction'
@derivative(of: nonexistentFunction)
                ^
```

Example:

```swift
import _Differentiation

func original(_ x: Int) -> Int { x }

@derivative(of: original)
func vjpOriginalFunctionNotFound2(_ x: Float) -> (value: Int, pullback: (Float) -> Float) {
  fatalError()
}
```

Before: unideal diagnostic about `value:` result.
```
derivative2.swift:5:2: error: '@derivative(of:)' attribute requires function to return a two-element tuple (first element type 'Int' must conform to 'Differentiable')
@derivative(of: original)
 ^
```

After: ideal diagnostic about original function not found.
```
derivative2.swift:5:17: error: could not find function 'original' with expected type '(Float) -> Int'
@derivative(of: original)
                ^
```

---

Let me know if these changes make sense, and if you'd like further changes! I'd like to defer major diagnostic changes until later to unblock progress.

Reply:
This is very nice!